examples/prom/nonlinear_elasticity_global_rom.cpp

```diff
@@ -0,0 +1,1747 @@
+{
+    const int dim = x.Size();
+    const double s_eff = s / 80.0;
+
+    v = 0.0;
+    v(dim - 1) = -s_eff * x(0);
+}
+
+void InitialDeformationIC2(const Vector &x, Vector &y) //See MFEM ex19
+{
+    // Set the initial configuration. Having this different from the reference
+    // configuration can help convergence
+    const int dim = x.Size();
+    const double s_eff = s;
+    y = x;
+    y(dim-1) = x(dim - 1) + 0.25*x(0) * s_eff;
+}
+
+void InitialVelocityIC2(const Vector& x, Vector& v)
+{
+    v = 0.0;
+}
+
+RomOperator::RomOperator(HyperelasticOperator* fom_,
+                         HyperelasticOperator* fomSp_, const int rvdim_, const int rxdim_,
+                         const int hdim_, CAROM::SampleMeshManager* smm_, const Vector v0_,
+                         const Vector x0_, const Vector v0_fom_, const CAROM::Matrix* V_v_,
+                         const CAROM::Matrix* V_x_, const CAROM::Matrix* U_H_,
+                         const CAROM::Matrix* Hsinv_, const int myid, const bool oversampling_,
+                         const bool hyperreduce_, const bool x_base_only_)
+    : TimeDependentOperator(rxdim_ + rvdim_, 0.0), fom(fom_), fomSp(fomSp_),
+      rxdim(rxdim_), rvdim(rvdim_), hdim(hdim_), x0(x0_), v0(v0_), v0_fom(v0_fom_),
+      smm(smm_), nsamp_H(smm_->GetNumVarSamples("H")), V_x(*V_x_), V_v(*V_v_),
+      U_H(U_H_), Hsinv(Hsinv_), zN(std::max(nsamp_H, 1), false), zX(std::max(nsamp_H,
+              1), false), M_hat_solver(fom_->fespace.GetComm()), oversampling(oversampling_),
+      z(height / 2), hyperreduce(hyperreduce_), x_base_only(x_base_only_)
+{
+    if (myid == 0)
+    {
+        V_v_sp = new CAROM::Matrix(fomSp->Height() / 2, rvdim, false);
+        V_x_sp = new CAROM::Matrix(fomSp->Height() / 2, rxdim, false);
+    }
+
+    // Gather distributed vectors
+    if (x_base_only)
+    {
+        smm->GatherDistributedMatrixRows("X", V_v, rvdim, *V_v_sp);
+    }
+    else
+    {
+        smm->GatherDistributedMatrixRows("V", V_v, rvdim, *V_v_sp);
+    }
+
+    smm->GatherDistributedMatrixRows("X", V_x, rxdim, *V_x_sp);
+
+    // Create V_vTU_H, for hyperreduction
+    V_v.transposeMult(*U_H, V_vTU_H);
+
+    S_hat = new CAROM::Matrix(rvdim, rvdim, false);
+    S_hat_v0 = new CAROM::Vector(rvdim, false);
+    S_hat_v0_temp = new Vector(v0_fom.Size());
+    S_hat_v0_temp_librom = new CAROM::Vector(S_hat_v0_temp->GetData(),
+            S_hat_v0_temp->Size(), true, false);
+    M_hat = new CAROM::Matrix(rvdim, rvdim, false);
+    M_hat_inv = new CAROM::Matrix(rvdim, rvdim, false);
+
+    // Create S_hat
+    Compute_CtAB(&(fom->Smat), V_v, V_v, S_hat);
+
+    // Apply S_hat to the initial velocity and store
+    fom->Smat.Mult(v0_fom, *S_hat_v0_temp);
+    V_v.transposeMult(*S_hat_v0_temp_librom, S_hat_v0);
+
+    // Create M_hat
+    Compute_CtAB(fom->Mmat, V_v, V_v, M_hat);
+
+    // Invert M_hat and store
+    M_hat->inverse(*M_hat_inv);
+
+    //if (myid == 0)
```

Comment:
Should `myid == 0` be used or removed? Has this example been tested in parallel?

Reply:
@axla-io I have just tested it in parallel and found that it crashes.

Reply:
@chldkdtn @dylan-copeland That is disturbing. I think that I tried it in parallel with my two processors, but I'm not sure. I will try it on my side and do some debugging.

I'll also check whether the `myid == 0` should be there.

Reply:
@chldkdtn Just to be clear, does the `offline` and `merge` modes work for you? They do for me, but I get a crash on the `online` mode, which I'll debug.

Reply:
@chldkdtn @dylan-copeland 
I've fixed the MPI issue for the online mode in the latest commit. It seems, however, that I get into a race condition when I use MPI for the merge mode with more than two snapshots (for parametric ROM). This is surprising to me as I followed the `mixed_nonlinear_diffusion` code closely. Could you take a look and see if you also get a race condition with the latest commit if you run:

```
mpirun -np 2 nonlinear_elasticity_global_rom --offline -dt 0.01 -tf 5.0 -s 14 -vs 100 -sc 0.90 -xbo -def-ic -id 0

mpirun -np 2 nonlinear_elasticity_global_rom --offline -dt 0.01 -tf 5.0 -s 14 -vs 100 -sc 1.10 -xbo -def-ic -id 1

mpirun -np 2 nonlinear_elasticity_global_rom --merge -ns 2 -dt 0.01 -tf 5.0 -xbo
```

Oddly enough, running
```
mpirun -np 2 nonlinear_elasticity_global_rom --merge -ns 1 -dt 0.01 -tf 5.0 -xbo
```

Works for me, so maybe it's just that my computer is slow? It could also be that the `MergeBasis` function from `mixed_nonlinear_diffusion` is not made to handle parametric ROMs, but that's just speculation.

Reply:
Also, it turns out that there is some difference for the hyper reduced results between parallel and serial mode. I think that we need to understand whether this is connected to the merge issue as well before looking into it more.

Reply:
The `MergeBasis` function in `mixed_nonlinear_diffusion` works for parametric ROM in serial. There is a serial sample run that works, to verify this. There may be multiple reasons why that test does not work in parallel, and I'm not if it involves the merge phase. I ran the above sample run for `nonlinear_elasticity_global_rom` and also found that it hangs. At first glance, I don't see what the issue is. If it is difficult to debug, I would say that getting this PR to work in serial is good enough, since other examples in the `prom` directory do not work in parallel yet.

Reply:
I understand. For the parametric ROM, I should have precised that I meant that the `mixed_nonlinear_diffusion` example maybe couldn't handle parametric ROMs in parallel.

But then, for now this is good enough now and more effort can be put into the parallelization at a later stage? Maybe making the snapshot writing, and base merging serial could solve our issues?

Reply:
In general, snapshot writing and basis merging should be parallel. Laghos does these things in parallel. It is important to keep the data distributed, since in theory large problems could have so much data that memory limits are exceeded in serial, not to mention the performance bottleneck. So it should work in parallel, but it has to be debugged. For now, I am fine with leaving this for future work.

Reply:
Thanks Dylan, that makes a lot of sense. I guess there is also a bottleneck in the implementation here. In `mixed_linear_diffusion`, the ROM operator only works in serial in the online mode, same goes for this program. So in order to have parallelization we need to change the implementation of the ROM online mode as well.

Reply:
@axla-io is it possible to make the parallel version work in this PR? That would be wonderful.

Reply:
@axla-io The fact that `mixed_nonlinear_diffusion` only works in serial is due to bugs, and the question of whether the online stage should be serial or parallel is separate but important. In that example, you can see that `ode_solver.Step` is called in the online case only for rank 0. The idea is that the online stage must be run in parallel in order to load the parallel basis files (created by the merge stage), but the ROM problem is too small to parallelize. Ideally, we would have another stage as in Laghos for preprocessing in parallel everything that is needed for the serial online stage and writing it to file. For example, the preprocessing stage would load the basis files in parallel and compute the reduced matrices and write them to file, to be read by the serial online stage. That level of optimization is not really necessary in these examples.

Also, to clarify, I think it would be acceptable to have this example work only in serial for this PR, but of course it is worth trying to fix the parallel case if it doesn't take too much work.

Reply:
@axla-io Let's not worry about parallel implementation in this PR as @dylan-copeland said.

Reply:
Thanks for the explanation @dylan-copeland !

Reply:
@chldkdtn Okay, that sounds good. I'll turn my focus to the report writing now.